pkg/kubelet/prober/worker.go

```diff
@@ -0,0 +1,156 @@
+	pod *api.Pod
+
+	// The container to probe (read-only)
+	container api.Container
+
+	// Describes the probe configuration (read-only)
+	spec *api.Probe
+
+	// The last known container ID for this worker.
+	containerID types.UID
+}
+
+// Creates and starts a new probe worker.
+func (m *manager) newWorker(
+	pod *api.Pod,
+	container api.Container) *worker {
+
+	w := &worker{
+		stop:      make(chan struct{}),
+		pod:       pod,
+		container: container,
+		spec:      container.ReadinessProbe,
+	}
+
+	// Start the worker thread.
+	go run(m, w)
+
+	return w
+}
+
+// run periodically probes the container.
+func run(m *manager, w *worker) {
+	probeTicker := time.Tick(m.defaultProbePeriod)
+
+probeLoop:
+	for doProbe(m, w) {
+		// Wait for next probe tick.
+		select {
+		case <-w.stop:
+			break probeLoop
+		case <-probeTicker:
+			// continue
+		}
+	}
+
+	// Clean up.
+	if w.containerID != "" {
+		m.readinessCache.removeReadiness(string(w.containerID))
+	}
+
+	m.workerLock.Lock()
+	defer m.workerLock.Unlock()
+	delete(m.readinessProbes, containerPath{w.pod.UID, w.container.Name})
+}
+
+// doProbe probes the container once and records the result.
+// Returns whether the worker should continue.
+func doProbe(m *manager, w *worker) bool {
+	status, ok := m.statusManager.GetPodStatus(w.pod.UID)
+	if !ok {
+		// Either the pod has not been created yet, or it was already deleted.
+		glog.V(3).Infof("No status for pod: %v", kubecontainer.GetPodFullName(w.pod))
+		status = w.pod.Status // Fallback to status from apiserver
+	}
+
+	// Worker should terminate if pod is terminated.
+	if status.Phase == api.PodFailed || status.Phase == api.PodSucceeded {
+		glog.V(3).Infof("Pod %v %v, exiting probe worker",
+			kubecontainer.GetPodFullName(w.pod), status.Phase)
+		return false
+	}
+
+	c, ok := api.GetContainerStatus(status.ContainerStatuses, w.container.Name)
+	if !ok {
+		// Either the container has not been created yet, or it was deleted.
+		glog.V(3).Infof("Non-existant container probed: %v - %v",
+			kubecontainer.GetPodFullName(w.pod), w.container.Name)
+		return true // Wait for more information.
+	}
+
+	if w.containerID != types.UID(c.ContainerID) {
+		if w.containerID != "" {
+			m.readinessCache.removeReadiness(string(w.containerID))
+		}
+		w.containerID = types.UID(c.ContainerID)
+	}
+
+	if c.State.Running == nil {
+		glog.V(3).Infof("Non-running container probed: %v - %v",
+			kubecontainer.GetPodFullName(w.pod), w.container.Name)
+		m.readinessCache.setReadiness(string(w.containerID), false)
+		// Abort if the container will not be restarted.
+		return c.State.Terminated == nil ||
+			w.pod.Spec.RestartPolicy != api.RestartPolicyNever
+	}
+
+	remainingDelay := w.spec.InitialDelaySeconds - int64(time.Since(c.State.Running.StartedAt.Time).Seconds())
+	if remainingDelay > 0 {
+		// Readiness defaults to false during the initial delay.
+		m.readinessCache.setReadiness(string(w.containerID), false)
+
+		select {
```

Comment:
I understand that this is here so that the worker can wake up at the right time (after the initial delay), but IMO this makes the control flow unnecessarily complicated. Suggestion: remove this `select` and simply return true if the initial delay has not passed yet.